docs/convenience.rst

```diff
@@ -63,119 +61,64 @@ overcome transient failures.  This typically looks something like:
         else:
            break
 
-This is rather ugly.
+This is rather ugly and easy to get wrong.
 
-Transaction managers provide a helper for this case. To show this,
-we'll use a contrived example:
+Transaction managers provide two helpers for this case.
 
-.. doctest::
+Running and retrying functions as transactions
+______________________________________________
 
-    >>> ntry = 0
-    >>> with transaction.manager:
-    ...      dm['ntry'] = 0
-
-    >>> import transaction.interfaces
-    >>> class Retry(transaction.interfaces.TransientError):
-    ...     pass
-
-    >>> for attempt in transaction.manager.attempts():
-    ...     with attempt as t:
-    ...         t.note('test')
-    ...         print("%s %s" % (dm['ntry'], ntry))
-    ...         ntry += 1
-    ...         dm['ntry'] = ntry
-    ...         if ntry % 3:
-    ...             raise Retry(ntry)
-    0 0
-    0 1
-    0 2
-
-The raising of a subclass of TransientError is critical here. It's
-what signals that the transaction should be retried.  It is generally
-up to the data manager to signal that a transaction should try again
-by raising a subclass of TransientError (or TransientError itself, of
-course).
```

Comment:
It's may also be worth noting that a doomed transaction will stop the entire process too. A DoomedTransaction exception isn't transient, and will be raised from `txn.commit()` once `txn.doom()` is called, but putting all those pieces together isn't necessarily obvious. And one could easily imagine that `run` has special handling for `DoomedTransaction`.